bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/services/UrlProviderImpl.java

```diff
@@ -53,38 +56,77 @@
 
     private String productUrlTemplate;
     private Pair<IdentifierLocation, ProductIdentifierType> productIdentifierConfig;
 
     private String categoryUrlTemplate;
-    private Pair<IdentifierLocation, CategoryIdentifierType> categoryIdentifierConfig;
+    private IdentifierLocation categoryIdentifierLocation;
+
+    private String queryParameterName;
 
     @Activate
     public void activate(UrlProviderConfiguration conf) {
         productUrlTemplate = conf.productUrlTemplate();
         productIdentifierConfig = Pair.of(conf.productIdentifierLocation(), conf.productIdentifierType());
 
         categoryUrlTemplate = conf.categoryUrlTemplate();
-        categoryIdentifierConfig = Pair.of(conf.categoryIdentifierLocation(), conf.categoryIdentifierType());
+        categoryIdentifierLocation = conf.categoryIdentifierLocation();
+
+        queryParameterName = conf.identifierQueryParamater();
     }
 
     @Override
     public String toProductUrl(SlingHttpServletRequest request, Page page, Map<String, String> params) {
         return toUrl(request, page, params, productUrlTemplate);
     }
 
+    @Override
+    public String toProductUrl(SlingHttpServletRequest request, Page page, String productIdentifier, MagentoGraphqlClient graphqlClient) {
+        ParamsBuilder params = new ParamsBuilder();
+        if (graphqlClient != null && StringUtils.isNotBlank(productIdentifier)) {
+            params.sku(productIdentifier);
+            ProductUrlParameterRetriever retriever = new ProductUrlParameterRetriever(graphqlClient);
+            retriever.setIdentifier(productIdentifier);
+            ProductInterface product = retriever.fetchProduct();
+            if (product != null) {
+                params.urlKey(product.getUrlKey());
+            } else {
+                LOGGER.debug("Could not generate product page URL for {}.", productIdentifier);
+            }
+        }
+        return toUrl(request, page, params.map(), productUrlTemplate);
+    }
+
     @Override
     public String toCategoryUrl(SlingHttpServletRequest request, Page page, Map<String, String> params) {
         return toUrl(request, page, params, categoryUrlTemplate);
     }
 
+    @Override
+    public String toCategoryUrl(SlingHttpServletRequest request, Page page, String categoryIdentifier, MagentoGraphqlClient graphqlClient) {
+        ParamsBuilder params = new ParamsBuilder().uid(categoryIdentifier);
+        if (graphqlClient != null && StringUtils.isNotBlank(categoryIdentifier)) {
+            CategoryUrlParameterRetriever retriever = new CategoryUrlParameterRetriever(graphqlClient);
+            retriever.setIdentifier(categoryIdentifier);
+            CategoryInterface category = retriever.fetchCategory();
+            if (category != null) {
+                params.urlKey(category.getUrlKey()).urlPath(category.getUrlPath());
+            } else {
+                LOGGER.debug("Could not generate category page URL for {}.", categoryIdentifier);
```

Comment:
Is a fall-through the right choice here? How would a URL to a category that cannot be fetched look like?

Reply:
Same as above, if URL parameters are missing we cannot do much.